File: 2017/day-17-swift/PartOne.swift
// http://adventofcode.com/2017/day/17

func solve(steps: Int) -> Int {
  var buffer = [0]
  var index = 0

  for value in 1...2017 {
    index = (index + steps) % buffer.count + 1
    buffer.insert(value, at: index)
  }

  return buffer[index + 1]
}

File: 2017/day-05-go/part_one.go
// http://adventofcode.com/2017/day/5

package main

import "fmt"

type Program struct {
  steps int
  index int
  jumps []int
}

func Exec(program Program) Program {
  jump := program.jumps[program.index]
  program.jumps[program.index] += 1

  return Program {
    steps: program.steps + 1,
    index: program.index + jump,
    jumps: program.jumps,
  }
}

func IsRunning(program Program) bool {
  return program.index >= 0 && program.index < len(program.jumps)
}

func Solve([]int jumps) {
  program := Program { jumps: jumps }

  for IsRunning(program) {
    program = Exec(program)
  }

  return program.steps
}
